cmd/snap/cmd_info.go

```diff
@@ -163,27 +164,78 @@ func coalesce(snaps ...*client.Snap) *client.Snap {
 	return nil
 }
 
-// formatDescr formats a given string (typically a snap description)
+// runesTrimRightSpace returns text, with any trailing whitespace dropped.
+func runesTrimRightSpace(text []rune) []rune {
+	j := len(text)
+	for j > 0 && unicode.IsSpace(text[j-1]) {
+		j--
+	}
+	return text[:j]
+}
+
+// runesLastIndexSpace returns the index of the last whitespace rune
+// in the text. If the text has no whitespace, returns -1.
+func runesLastIndexSpace(text []rune) int {
+	for i := len(text) - 1; i >= 0; i-- {
+		if unicode.IsSpace(text[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// wrapLine wraps a line to fit into width, preserving the line's indent, and
+// writes it out prepending padding to each line.
+func wrapLine(out io.Writer, text []rune, pad string, width int) {
+	// Note: this is _wrong_ for much of unicode (because the width of a
+	//       rune on the terminal is anything between 0 and 2, not always 1
+	//       as this code assumes) but fixing that is Hard. Long story
+	//       short, you can get close using a couple of big unicode tables
+	//       (which is what wcwidth does). Getting it 100% requires a
+	//       terminfo-alike of unicode behaviour.
+	// This (and possibly printDescr below) should move to strutil once
+	// we're happy with it getting wider (heh heh) use.
+
+	// discard any trailing whitespace
+	text = runesTrimRightSpace(text)
+	// establish the indent of the whole block
+	idx := 0
+	for idx < len(text) && unicode.IsSpace(text[idx]) {
+		idx++
+	}
+	indent := pad + string(text[:idx])
```

Comment:
Looking through this function, is there an actual benefit in using []rune instead of just string? It looks like it's trimming spaces from the left side, which is achievable with strings.TrimRightFunc, and then picking the last space out of it to split, which is doable with strings.IndexLastFunc. The indent and the padding are already being handled as plain strings, and the output is also back into a raw buffer. So it's not clear there's much benefit in the []rune conversions, for clarity or otherwise.

Reply:
the main benefit is that the length of the string is not the space it'll take up on the terminal; the rune length is a much better approximation (where if we're doing successive approximations, the next step is to use something like wcwidth, i.e. a couple of table lookups more per rune). 

The width of the indent and the padding is counted in runes. Which, sure, can be done with `RuneCountInString`, as I do for the pad.

Finding the last whitespace shorter than width is the hard one, because width is the rune count. You can't just slice the string, as you don't even know if you're slicing a rune in half.